### DISCUSSION TITLE
How to use sampling on time-series table sorted by DateTime

### DISCUSSION THREAD

**Discussion by eKrajnak** [ClickHouse/ClickHouse#68635] [Q&A] [answered]
Hello,

I use MergeTree table to store metadata of sampled IP packets (thousands per second). I need to perform many different searches on that dataset and keep the queries fast at the same time. Because of that I want to use clickhouse sampling when displaying wide time window (24h time-series graph need to load n x 150M rows).

I had a few ideas, but:

- when using primary key (timestamp) and sample (timestamp) - sampling don't work because of low randomness of timestamp
- when using primary key (timestamp, cityHash64(uuid)) and sample (cityHash64(uuid)) - sampling don't working because timestamp is high cardinality column
- when using primary key (cityHash64(uuid), timestamp) and sample (cityHash64(uuid)) - sampling works, displaying wide time windows is fast, but displaying very short time window is too slow (because of order of primary key clickhouse can not filter on timestamp column)

Please, how to achieve all the requirements mentioned above? Is there any way to sample on DateTime column without the need of using other randomness (uuid) and hashing?

Thanks
Blažej

**Comment by den-crane** [comment, marked as answer]
```
partition by toDate(timestamp)
order by ( 
    toStartOfHour(timestamp),    ----<<<<<<<     toStartOfHour !
    cityHash64(uuid))
```

https://kb.altinity.com/altinity-kb-queries-and-syntax/sampling-example/

**Reply by eKrajnak** [reply]
> ```
> partition by toDate(timestamp)
> order by ( 
>     toStartOfHour(timestamp),    ----<<<<<<<     toStartOfHour !
>     cityHash64(uuid))
> ```
> 
> https://kb.altinity.com/altinity-kb-queries-and-syntax/sampling-example/

I tried this order by / primary key, but now I experience no usage of timestamp (first element in primary key). Even when I request very short time window, the CH still read the whole time range (actually 24 hours).

```
EXPLAIN indexes=1
SELECT src_ip
FROM "default"."data_new"
WHERE 
  (
  toStartOfInterval(toDateTime(time_received), INTERVAL 10 second) >= toDateTime(1724220711) AND 
  toStartOfInterval(toDateTime(time_received), INTERVAL 10 second) <= toDateTime(1724230915)
  ) AND src_ip = '8.8.8.8'
GROUP BY src_ip
ORDER BY sum(bytes) as bits DESC
LIMIT 10
```
```
    ┌─explain───────────────────────────────────────────────────┐
 1. │ Expression (Project names)                                │
 2. │   Limit (preliminary LIMIT (without OFFSET))              │
 3. │     Sorting (Sorting for ORDER BY)                        │
 4. │       Expression ((Before ORDER BY + Projection))         │
 5. │         Aggregating                                       │
 6. │           Expression (Before GROUP BY)                    │
 7. │             Expression                                    │
 8. │               ReadFromMergeTree (default.data_new)        │
 9. │               Indexes:                                    │
10. │                 PrimaryKey                                │
11. │                   Condition: true                         │
12. │                   Parts: 6/6                              │
13. │                   Granules: 307765/307765                 │
14. │                 Skip                                      │
15. │                   Name: src_ip                            │
16. │                   Description: bloom_filter GRANULARITY 1 │
17. │                   Parts: 6/6                              │
18. │                   Granules: 109674/307765                 │
    └───────────────────────────────────────────────────────────┘
```

**Reply by den-crane** [reply]
?  toStartOfInterval(toDateTime(), INTERVAL 10 second)

**Reply by eKrajnak** [reply]
It's generated statement by Grafana.
But I think I got the point. I must use 2 conditions: one which uses primary key and must be as "... toStartOfHours(time_received) ..." and the second for exact time match. Now it works.

**Reply by den-crane** [reply]
>It's generated statement by Grafana.

Then materialize toStartOfHour to a column time_received_h and use it
